core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -417,21 +421,54 @@ public void registerMovingAverageModel(ParseField modelName, Writeable.Reader<Mo
-     * @param internalReader reads the {@link InternalAggregation} from a stream
-     * @param aggregationParser reads the aggregation builder from XContent
-     * @param aggregationName names by which the aggregation may be parsed. The first name is special because it is the name that the reader
-     *        is registered under.
      */
-    public void registerAggregation(Writeable.Reader<? extends AggregationBuilder> builderReader,
-            Writeable.Reader<? extends InternalAggregation> internalReader, Aggregator.Parser aggregationParser,
-            ParseField aggregationName) {
+    public void registerAggregation(AggregationSpec spec) {
         if (false == transportClient) {
-            namedWriteableRegistry.register(AggregationBuilder.class, aggregationName.getPreferredName(), builderReader);
-            aggregationParserRegistry.register(aggregationParser, aggregationName);
+            namedWriteableRegistry.register(AggregationBuilder.class, spec.aggregationName.getPreferredName(), spec.builderReader);
+            aggregationParserRegistry.register(spec.aggregationParser, spec.aggregationName);
+        }
+        for (Map.Entry<String, Writeable.Reader<? extends InternalAggregation>> t : spec.internalReaders.entrySet()) {
+            String writeableName = t.getKey();
+            Writeable.Reader<? extends InternalAggregation> internalReader = t.getValue();
+            namedWriteableRegistry.register(InternalAggregation.class, writeableName, internalReader);
+        }
+    }
+
+    public static class AggregationSpec {
+        private final Map<String, Writeable.Reader<? extends InternalAggregation>> internalReaders = new TreeMap<>();
+        private final Writeable.Reader<? extends AggregationBuilder> builderReader;
+        private final Aggregator.Parser aggregationParser;
+        private final ParseField aggregationName;
+
+        /**
+         * Register an aggregation.
+         *
+         * @param builderReader reads the {@link AggregationBuilder} from a stream
+         * @param aggregationParser reads the aggregation builder from XContent
+         * @param aggregationName names by which the aggregation may be parsed. The first name is special because it is the name that the
+         *          reader is registered under.
+         */
+        public AggregationSpec(Reader<? extends AggregationBuilder> builderReader, Parser aggregationParser, ParseField aggregationName) {
+            this.builderReader = builderReader;
+            this.aggregationParser = aggregationParser;
+            this.aggregationName = aggregationName;
+        }
+
+        /**
+         * Add a reader for the shard level results of the aggregation with {@linkplain aggregationName}'s
+         * {@link ParseField#getPreferredName()} as the {@link NamedWriteable#getWriteableName()}.
+         */
+        public AggregationSpec addInternalReader(Writeable.Reader<? extends InternalAggregation> internalReader) {
```

Comment:
Can we call them `AggregationReader` or maybe `AggregationResultReader` since `InternalReader` doesn't really reflect what they are?